DICOM/src/Microsoft.Health.DeID.SharedLib/RedactFunction.cs

```diff
@@ -0,0 +1,166 @@
+{
+    public class RedactFunction
+    {
+        private static readonly int AgeThreshold = 89;
+        private static readonly string ReplacementDigit = "0";
+        private static readonly int InitialDigitsCount = 3;
+        private readonly RedactSetting _redactSetting;
+
+        public RedactFunction(RedactSetting redactSetting = null)
+        {
+            _redactSetting = redactSetting ?? new RedactSetting();
+        }
+
+        public string Redact(string inputString, AnonymizerValueTypes valueType)
+        {
+            return valueType switch
+            {
+                AnonymizerValueTypes.Date => RedactDate(inputString),
+                AnonymizerValueTypes.DateTime => RedactDateTime(inputString),
+                AnonymizerValueTypes.Age => RedactAge(decimal.Parse(inputString)).ToString(),
+                AnonymizerValueTypes.PostalCode => RedactPostalCode(inputString),
+                _ => null,
+            };
+        }
+
+        public uint? Redact(uint value, AnonymizerValueTypes valueType)
+        {
+            return valueType switch
+            {
+                AnonymizerValueTypes.Age => RedactAge(value),
+                _ => null,
+            };
+        }
+
+        public decimal? Redact(decimal value, AnonymizerValueTypes valueType)
+        {
+            return valueType switch
+            {
+                AnonymizerValueTypes.Age => RedactAge(value),
+                _ => null,
+            };
+        }
+
+        public DateTimeOffset? Radact(DateTimeOffset dateTime)
+        {
+            EnsureArg.IsNotNull<DateTimeOffset>(dateTime, nameof(dateTime));
+
+            if (_redactSetting.EnablePartialDatesForRedact)
+            {
+                if (DateTimeUtility.IndicateAgeOverThreshold(dateTime))
+                {
+                    return null;
+                }
+
+                return new DateTimeOffset(dateTime.Year, 1, 1, 0, 0, 0, default);
+            }
+
+            return null;
+        }
+
+        public DateTimeObject Redact(DateTimeObject dateObject)
+        {
+            EnsureArg.IsNotNull(dateObject, nameof(dateObject));
+
+            if (_redactSetting.EnablePartialDatesForRedact)
+            {
+                if (DateTimeUtility.IndicateAgeOverThreshold(dateObject.DateValue))
+                {
+                    return null;
+                }
+
+                dateObject.DateValue = new DateTimeOffset(dateObject.DateValue.Year, 1, 1, 0, 0, 0, dateObject.HasTimeZone == true ? dateObject.DateValue.Offset : new TimeSpan(0, 0, 0));
+                return dateObject;
+            }
+
+            return null;
+        }
+
+        public AgeObject Redact(AgeObject age)
+        {
+            EnsureArg.IsNotNull(age, nameof(age));
+
+            return RedactAge(age.AgeInYears()) == null ? null : age;
+        }
+
+        private string RedactDate(string dateString)
+        {
+            if (string.IsNullOrEmpty(dateString))
```

Comment:
[Discussion]

I would like to discuss the exception handling in this file. I found that in some functions, it return null simply when the input is not valid, but for some other functions, we use EnsureArg to check the parameter (and when it fails, it throws exception), so can we recheck the file and make the principals unified?

Reply:
Only for redact function, the input could be null. Other functions we use EnsureArg to check the parameters.